File: assn2/qsort.py
#!/usr/bin/python3

import random

def partition(array, low, high):
    index = low-1
    pivot = array[high]

    for i in range(low, high):
        if array[i] <= pivot:
            index = index + 1
            array[index], array[i] = array[i], array[index]
    
    array[index+1], array[high] = array[high], array[index+1]
    return(index+1)

def qsort(array, low, high):
    if low < high:
        newIndex = partition(array,low,high)
        qsort(array, low, newIndex-1)
        qsort(array, newIndex+1, high)

for i in range(10):
    arr = [0]*random.randint(10,20)

    for j in range(len(arr)):
        arr[j] = random.randint(0,100)

    print("Before Quick Sort:", arr)
    qsort(arr,0,len(arr)-1)
    print("After Quick sort:", arr, "\n\n")

File: assn2/qsort.cpp
#include <iostream>
#include <vector>
#include <stdlib.h>
#include <time.h>

using namespace std;

void swap(int* a, int* b){
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(int array[], int low, int high){
    int pivot = array[high];
    int index = low-1;
    int loop = high - low;

    for(int i=low; i<=high-1; i++){
        if(array[i] <= pivot){
            index++;
            swap(&array[index], &array[i]);
        }
    }
    swap(&array[index+1], &array[high]);
    return index+1;
}

void qsort(int array[], int low, int high){
    if(low<high){
        int newIndex = partition(array,low,high);
        qsort(array, low, newIndex-1);
        qsort(array, newIndex+1, high);
    }
}

int main() {
    srand(time(NULL));
    int numRange;

    for(int i=0; i<10; i++){
        numRange = rand() % 20 + 10;
        int arr[numRange];
        for(int j=0; j<numRange; j++){arr[j] = rand() % 100;}

        cout << "Before Quick Sort: \n[ ";
        for(int k=0; k<numRange; k++){cout << arr[k] << " ";}
        qsort(arr, 0, numRange-1);
        cout << "]\nAfter Quick Sort: \n[ ";
        for(int k=0; k<numRange; k++){cout << arr[k] << " ";}
        cout << "]\n\n";
    }

    return 0;
}
